### PR TITLE
Add new CLI option for --until

### PR DESCRIPTION
**Pull request by jamiehannaford**
**- What I did**

Added an --until flag for docker logs in line with #32807.

**- How I did it**

Adapted CLI pkg.

**- How to verify it**

Run unit/integration tests. One caveat is that we don't have any integration tests for any of the logger adaptors at the moment (e.g. loggerd), so I was not able to verify the changes to this.

**- Description for the changelog**
```
docker logs now supports --until
```

### CODE REVIEW

**Comment by thaJeztah** [comment]
design was reviewed in https://github.com/moby/moby/issues/32914, so moving this to code review

File: cli/command/container/logs.go [hunk 1]
```diff
@@ -3,9 +3,9 @@ package container
 import (
 	"io"
 
-	"github.com/docker/cli/cli"
-	"github.com/docker/cli/cli/command"
 	"github.com/docker/docker/api/types"
+	"github.com/docker/docker/cli"
+	"github.com/docker/docker/cli/command"
```
Comment (thaJeztah) [inline cli/command/container/logs.go]: Looks like these imports are changed by accident?

File: cli/command/container/logs.go [hunk 2]
```diff
@@ -73,4 +76,4 @@ func runLogs(dockerCli *command.DockerCli, opts *logsOptions) error {
 		_, err = stdcopy.StdCopy(dockerCli.Out(), dockerCli.Err(), responseBody)
 	}
 	return err
-}
+}
```
Comment (thaJeztah) [inline cli/command/container/logs.go]: Can you re-add the newline here?

File: cli/command/container/logs.go (hunk 1 above)
Comment (jamiehannaford) [inline cli/command/container/logs.go]: Yeah, I've just changed them back. I'll re-push what I have but the tests will stay broken.

**Comment by jamiehannaford** [comment]
Just as a FYI: this PR is contigent on moby/moby#32914 being merged first because it needs access to `github.com/docker/docker/api/types.ContainerLogsOptions`.

File: cli/command/container/logs.go [hunk 3]
```diff
@@ -38,6 +39,7 @@ func NewLogsCommand(dockerCli *command.DockerCli) *cobra.Command {
 	flags := cmd.Flags()
 	flags.BoolVarP(&opts.follow, "follow", "f", false, "Follow log output")
 	flags.StringVar(&opts.since, "since", "", "Show logs since timestamp (e.g. 2013-01-02T13:23:37) or relative (e.g. 42m for 42 minutes)")
+	flags.StringVar(&opts.until, "until", "", "Show logs before a timestamp (e.g. 2013-01-02T13:23:37) or relative (e.g. 42m for 42 minutes)")
```
Comment (thaJeztah) [inline cli/command/container/logs.go]: Can you add an API-version annotation on this flag? That makes sure the flag is automatically hidden if the client connects to an older daemon;

```go
flags.SetAnnotation("until", "version", []string{"1.30"})
```

**Comment by thaJeztah** [comment]
> Just as a FYI: this PR is contigent on moby/moby#32914 being merged first because it needs access to github.com/docker/docker/api/types.ContainerLogsOptions.

Thanks for the heads up 👍 I thought I'd give this an initial review already in parallel :smile:

**Comment by jamiehannaford** [comment]
@thaJeztah this is RFR again

File: vendor.conf [hunk 4]
```diff
@@ -47,7 +47,7 @@ github.com/xeipuuv/gojsonschema 93e72a773fade158921402d6a24c819b48aba29d
 golang.org/x/crypto 558b6879de74bc843225cde5686419267ff707ca
 golang.org/x/net 7dcfb8076726a3fdd9353b6b8a1f1b6be6811bd6
 golang.org/x/sync 450f422ab23cf9881c94e2db30cac0eb1b7cf80c
-golang.org/x/sys 07c182904dbd53199946ba614a412c61d3c548f5
+golang.org/x/sys 8dbc5d05d6edcc104950cc299a1ce6641235bc86
```
Comment (thaJeztah) [inline vendor.conf]: Perhaps this should use the same version as is used in moby/moby (95c6576299259db960f6c5b9b69ea52422860fce)

Comment (thaJeztah) [reply vendor.conf]: Did a quick check, and the following packages were updated in moby (so possibly should be updated here as well)

- https://github.com/golang/sys/compare/07c182904dbd53199946ba614a412c61d3c548f5...95c6576299259db960f6c5b9b69ea52422860fce
- https://github.com/containerd/continuity/compare/22694c680ee48fb8f50015b44618517e2bde77e8...35d55c5e8dd23b32037d56cf97174aff3efdfa83
- https://github.com/docker/swarmkit/compare/872861d2ae46958af7ead1d5fffb092c73afbaf0...28f91d87bd3f75fd039dbb9be49bfd2381019261

**Comment by jamiehannaford** [comment]
@thaJeztah Codecov is complaining about relative coverage? I added a whole new test case but it's not passing. What do you recommend?

**Comment by jamiehannaford** [comment]
@thaJeztah Does this LGTY?

**Comment by thaJeztah** [comment]
Thanks! Code changes LGTM; 

moving this to docs review: can you update the command-line reference, and:

- add the flag to the usage output; https://github.com/docker/cli/blob/master/docs/reference/commandline/logs.md#logs
- add an "examples" section, and a simple example (refer to https://github.com/docker/cli/blob/master/docs/reference/commandline/run.md#examples to see what an examples section could look like)
- copy the content of the examples section you created to a file named "logs-example.md" in the man/container/src/ directory: https://github.com/docker/cli/tree/master/man/src/container (that way the example is also included in the man pages for docker logs.

I notice the "description" section of the `docker logs` documentation needs a bit of cleaning up (and rewrite part of that to examples), but no need to do that as part of this pull request (I could take care of that as a follow up)

**Comment by jamiehannaford** [comment]
@thaJeztah Docs added :)

**Comment by thaJeztah** [comment]
Thanks! Erm, looks like there's something failing in CI;

```
cli/command/container/logs_test.go:60:125:warning: invalid operation: tc (variable of type struct{name string; args []string; expectedError string; containerInspectFunc func(img string) (github.com/docker/cli/vendor/github.com/docker/docker/api/types.ContainerJSON, error); containerLogsFunc func(string, github.com/docker/cli/vendor/github.com/docker/docker/api/types.ContainerLogsOptions) (io.ReadCloser, error)}) has no field or method opts (gosimple)
cli/command/container/hijack.go:188:1:warning: nolint directive did not match any issue (nolint)
cli/command/container/opts_test.go:46:1:warning: nolint directive did not match any issue (nolint)
cli/command/formatter/reflect_test.go:15:34:warning: nolint directive did not match any issue (nolint)
cli/command/container/logs_test.go:60:125:warning: parameter invalid operation: tc (variable of type struct{name string; args []string; expectedError string; containerInspectFunc func(img string) (github.com/docker/cli/vendor/github.com/docker/docker/api/types.ContainerJSON, error); containerLogsFunc func(string, github.com/docker/cli/vendor/github.com/docker/docker/api/types.ContainerLogsOptions) (io.ReadCloser, error)}) has no field or method opts (unparam)
cli/command/container/logs_test.go:60:125:warning: invalid operation: tc (variable of type struct{name string; args []string; expectedError string; containerInspectFunc func(img string) (github.com/docker/cli/vendor/github.com/docker/docker/api/types.ContainerJSON, error); containerLogsFunc func(string, github.com/docker/cli/vendor/github.com/docker/docker/api/types.ContainerLogsOptions) (io.ReadCloser, error)}) has no field or method opts (unused)
```


```
cli/command/container/logs_test.go:60: tc.opts undefined (type struct { name string; args []string; expectedError string; containerInspectFunc func(string) ("github.com/docker/cli/vendor/github.com/docker/docker/api/types".ContainerJSON, error); containerLogsFunc func(string, "github.com/docker/cli/vendor/github.com/docker/docker/api/types".ContainerLogsOptions) (io.ReadCloser, error) } has no field or method opts)
cli/command/container/logs_test.go:60: too many arguments in call to NewLogsCommand
	have (*test.FakeCli, <T>)
	want (command.Cli)
```

**Comment by jamiehannaford** [comment]
@thaJeztah Tests seem to be okay now

**Pull request merged**